euphoria-flink/src/main/java/cz/seznam/euphoria/flink/streaming/windowing/WindowAssigner.java

```diff
@@ -0,0 +1,105 @@
+/**
+ * Copyright 2016 Seznam.cz, a.s.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package cz.seznam.euphoria.flink.streaming.windowing;
+
+import cz.seznam.euphoria.core.client.dataset.windowing.Window;
+import cz.seznam.euphoria.core.client.dataset.windowing.WindowedElement;
+import cz.seznam.euphoria.core.client.dataset.windowing.Windowing;
+import cz.seznam.euphoria.core.client.functional.UnaryFunction;
+import cz.seznam.euphoria.flink.streaming.StreamingElement;
+import org.apache.flink.streaming.runtime.streamrecord.StreamRecord;
+
+import java.io.Serializable;
+import java.util.Set;
+import java.util.function.Function;
+
+/**
+ * Assigns windows to element and extracts key and value.
+ *
+ * @param <I>     type of input data
+ * @param <KEY>   type of output key
+ * @param <VALUE> type of output value
+ * @param <W>     type of output window
+ */
+public class WindowAssigner<I, KEY, VALUE, W extends Window>
+        implements Function<StreamRecord<StreamingElement<?, I>>, KeyedMultiWindowedElement<W, KEY, VALUE>>,
+        Serializable {
+
+  private final Windowing windowing;
+  private final UnaryFunction keyExtractor;
+  private final UnaryFunction valueExtractor;
+
+  private transient TimestampedElement reuse;
+
+  public WindowAssigner(Windowing windowing,
+                        UnaryFunction keyExtractor,
+                        UnaryFunction valueExtractor) {
+    this.windowing = windowing;
+    this.keyExtractor = keyExtractor;
+    this.valueExtractor = valueExtractor;
+  }
+
+  @Override
+  @SuppressWarnings("unchecked")
+  public KeyedMultiWindowedElement<W, KEY, VALUE> apply(StreamRecord<StreamingElement<?, I>> record) {
+    StreamingElement el = record.getValue();
+
+    if (reuse == null) {
+      reuse = new TimestampedElement();
+    }
+    reuse.setTimestamp(record.getTimestamp());
+    reuse.setElement(el);
+    Set windows = windowing.assignWindowsToElement(reuse);
```

Comment:
i wonder if we shouldn't call `reuse.setElement(null)` after the call to `assignWindowsToElement` to avoid referencing the last processed streaming element. but it probably makes no difference.